File: frontend/src/types/http-probes.ts
// HTTP Probes types that match the backend Pydantic schemas (backend/app/schemas/http_probes.py)

/**
 * HTTP Probe Record
 * 
 * Represents a single HTTP probe result from the database.
 * Matches the backend HTTPProbeResponse schema exactly.
 */
export interface HTTPProbe {
  id: string;                       // UUID
  scan_job_id: string;              // UUID of the scan job
  asset_id: string;                 // UUID of the asset
  url: string;                      // Full URL (e.g., "https://api.example.com")
  subdomain: string;                // Subdomain or IP (e.g., "api.example.com")
  parent_domain: string;            // Parent domain (e.g., "example.com")
  scheme: string;                   // "http" or "https"
  port: number;                     // Port number (e.g., 443, 80)
  status_code: number | null;       // HTTP status code (e.g., 200, 404)
  title: string | null;             // Page title
  webserver: string | null;         // Web server header (e.g., "nginx/1.18.0")
  content_length: number | null;    // Response content length in bytes
  content_type: string | null;      // Content-Type header (e.g., "text/html")
  final_url: string | null;         // Final URL after redirects
  ip: string | null;                // Resolved IP address
  technologies: string[];           // Detected technologies (e.g., ["React", "Webpack"])
  cdn_name: string | null;          // CDN provider if detected
  asn: string | null;               // Autonomous System Number
  chain_status_codes: number[];     // HTTP status codes in redirect chain
  location: string | null;          // Location header from redirects
  favicon_md5: string | null;       // MD5 hash of favicon
  created_at: string;               // ISO timestamp (e.g., "2025-11-16T17:33:03Z")
}

/**
 * HTTP Probe Statistics Response
 * 
 * Aggregate statistics for HTTP probes.
 * Returned by GET /api/v1/http-probes/stats/summary
 */
export interface HTTPProbeStats {
  total_probes: number;                           // Total number of probes
  status_code_distribution: Record<number, number>; // { 200: 13, 404: 4 }
  top_technologies: TechnologyCount[];            // Top detected technologies
  top_servers: ServerCount[];                     // Top web servers
  cdn_usage: Record<string, number>;              // CDN distribution
  redirect_chains_count: number;                  // Number of probes with redirects
}

/**
 * Technology Count
 * 
 * Used in top technologies list
 */
export interface TechnologyCount {
  name: string;   // Technology name (e.g., "React")
  count: number;  // Number of occurrences
}

/**
 * Server Count
 * 
 * Used in top servers list
 */
export interface ServerCount {
  name: string;   // Server name (e.g., "nginx/1.18.0")
  count: number;  // Number of occurrences
}

/**
 * HTTP Probe Query Parameters
 * 
 * Query parameters for fetching HTTP probes with filters and pagination.
 */
export interface HTTPProbeQueryParams {
  // Pagination
  limit?: number;                 // Default: 100, Max: 1000
  offset?: number;                // Default: 0
  
  // Filters
  asset_id?: string;              // Filter by asset UUID
  scan_job_id?: string;           // Filter by scan job UUID
  status_code?: number;           // Filter by HTTP status code
  scheme?: 'http' | 'https';      // Filter by scheme
  technology?: string;            // Filter by technology (e.g., "React")
  has_redirect?: boolean;         // Filter probes with redirect chains
  cdn_name?: string;              // Filter by CDN provider
}

/**
 * HTTP Probe List Response
 * 
 * Response from GET /api/v1/http-probes
 */
export type HTTPProbeListResponse = HTTPProbe[];

/**
 * HTTP Probe Cache Entry
 * 
 * Client-side cache entry for HTTP probes with TTL.
 */
export interface HTTPProbeCacheEntry {
  data: HTTPProbeListResponse;
  timestamp: number;              // Unix timestamp in milliseconds
  params: HTTPProbeQueryParams;   // Query params used for this cache entry
}

/**
 * HTTP Probe Fetch Options
 * 
 * Options for fetching HTTP probes with caching control.
 */
export interface HTTPProbeFetchOptions {
  forceRefresh?: boolean;         // Bypass cache
  signal?: AbortSignal;           // For request cancellation
}

/**
 * Status Code Categories
 * 
 * Helper for categorizing HTTP status codes by type.
 */
export const STATUS_CODE_CATEGORIES = {
  SUCCESS: [200, 201, 202, 203, 204],
  REDIRECT: [301, 302, 303, 307, 308],
  CLIENT_ERROR: [400, 401, 403, 404, 405, 429],
  SERVER_ERROR: [500, 502, 503, 504],
};

/**
 * Status Code Category Type
 */
export type StatusCodeCategory = keyof typeof STATUS_CODE_CATEGORIES;

/**
 * Helper function to get status code category
 */
export function getStatusCodeCategory(statusCode: number): StatusCodeCategory | null {
  if (STATUS_CODE_CATEGORIES.SUCCESS.includes(statusCode)) return 'SUCCESS';
  if (STATUS_CODE_CATEGORIES.REDIRECT.includes(statusCode)) return 'REDIRECT';
  if (STATUS_CODE_CATEGORIES.CLIENT_ERROR.includes(statusCode)) return 'CLIENT_ERROR';
  if (STATUS_CODE_CATEGORIES.SERVER_ERROR.includes(statusCode)) return 'SERVER_ERROR';
  return null;
}

/**
 * Helper function to check if probe has redirect chain
 */
export function hasRedirectChain(probe: HTTPProbe): boolean {
  return probe.chain_status_codes.length > 0;
}

/**
 * Helper function to check if probe is successful (2xx status)
 */
export function isSuccessfulProbe(probe: HTTPProbe): boolean {
  return probe.status_code !== null && probe.status_code >= 200 && probe.status_code < 300;
}

File: backend/app/schemas/http_probes.py
"""
HTTP Probe Schemas for API responses and validation.

This module defines Pydantic models for HTTP probe data from the httpx module,
enabling type-safe API interactions and database serialization.

Author: Pluckware Development Team
Date: November 17, 2025
Phase: HTTPx Frontend Implementation - Phase 1
"""

from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field


class HTTPProbeBase(BaseModel):
    """
    Base HTTP probe model with all fields from http_probes table.
    
    Represents a single HTTP probe result from the httpx module.
    Handles nullable fields and JSONB arrays based on real production data.
    """
    # Core URL fields (always present)
    url: str = Field(..., description="Full URL probed (e.g., https://autodiscover.greatlakescheese.com)")
    subdomain: str = Field(..., description="Fully qualified subdomain")
    parent_domain: str = Field(..., description="Parent/apex domain")
    scheme: str = Field(..., description="URL scheme: http or https")
    port: int = Field(..., description="Port number (443, 80, 8080, etc.)")
    
    # HTTP Response fields (nullable based on real data)
    status_code: Optional[int] = Field(None, description="HTTP status code (200, 404, 500, etc.)")
    title: Optional[str] = Field(None, description="HTML page title (can be null)")
    webserver: Optional[str] = Field(None, description="Server header value (e.g., 'Apache', 'Microsoft-IIS/10.0')")
    content_length: Optional[int] = Field(None, description="Response content length in bytes (can be null)")
    content_type: Optional[str] = Field(None, description="Content-Type header (e.g., 'text/html')")
    final_url: Optional[str] = Field(None, description="Final URL after following redirects")
    ip: Optional[str] = Field(None, description="Server IP address")
    
    # Technology detection (JSONB arrays - can be empty)
    technologies: List[str] = Field(
        default_factory=list, 
        description="Detected technologies with versions (e.g., ['IIS:10.0', 'Microsoft ASP.NET', 'Windows Server'])"
    )
    cdn_name: Optional[str] = Field(None, description="CDN provider name (e.g., 'Cloudflare', 'Akamai')")
    asn: Optional[str] = Field(None, description="Autonomous System Number")
    
    # Metadata fields
    chain_status_codes: List[int] = Field(
        default_factory=list,
        description="HTTP status codes from redirect chain (e.g., [302, 302, 200] or empty [])"
    )
    location: Optional[str] = Field(None, description="Location header from redirects")
    favicon_md5: Optional[str] = Field(None, description="MD5 hash of favicon (for clustering similar sites)")
    
    # Timestamp
    created_at: datetime = Field(..., description="Timestamp when probe was performed")
    
    class Config:
        from_attributes = True
        json_schema_extra = {
            "example": {
                "url": "https://autodiscover.greatlakescheese.com",
                "subdomain": "autodiscover.greatlakescheese.com",
                "parent_domain": "greatlakescheese.com",
                "scheme": "https",
                "port": 443,
                "status_code": 200,
                "title": "Outlook",
                "webserver": "Microsoft-IIS/10.0 Microsoft-HTTPAPI/2.0",
                "content_length": 58725,
                "content_type": "text/html",
                "final_url": "https://autodiscover.greatlakescheese.com/owa/auth/logon.aspx",
                "ip": "40.136.126.11",
                "technologies": ["IIS:10.0", "Microsoft ASP.NET", "Microsoft HTTPAPI:2.0", "Windows Server"],
                "cdn_name": None,
                "asn": None,
                "chain_status_codes": [302, 302, 200],
                "location": None,
                "favicon_md5": None,
                "created_at": "2025-11-16T17:33:03Z"
            }
        }


class HTTPProbeResponse(HTTPProbeBase):
    """
    HTTP probe response model for API responses.
    
    Extends HTTPProbeBase with identifier fields.
    Used when returning individual probe records via API endpoints.
    """
    id: str = Field(..., description="UUID of the HTTP probe record")
    scan_job_id: str = Field(..., description="Associated scan job UUID")
    asset_id: str = Field(..., description="Associated asset UUID")
    
    class Config:
        from_attributes = True
        json_schema_extra = {
            "example": {
                "id": "0462222e-963f-4936-adb2-1a75d53d8a9b",
                "scan_job_id": "302ef2b9-9bb5-4c06-a187-473b37904458",
                "asset_id": "17e70cea-9abd-4d4d-a71b-daa183e9b2de",
                "url": "https://autodiscover.greatlakescheese.com",
                "subdomain": "autodiscover.greatlakescheese.com",
                "parent_domain": "greatlakescheese.com",
                "scheme": "https",
                "port": 443,
                "status_code": 200,
                "title": "Outlook",
                "webserver": "Microsoft-IIS/10.0",
                "content_length": 58725,
                "content_type": "text/html",
                "final_url": "https://autodiscover.greatlakescheese.com/owa/auth/logon.aspx",
                "ip": "40.136.126.11",
                "technologies": ["IIS:10.0", "Microsoft ASP.NET", "Windows Server"],
                "cdn_name": None,
                "asn": None,
                "chain_status_codes": [302, 302, 200],
                "location": None,
                "favicon_md5": None,
                "created_at": "2025-11-16T17:33:03Z"
            }
        }


class HTTPProbeStatsResponse(BaseModel):
    """
    Aggregate statistics for HTTP probe results.
    
    Provides summary metrics for dashboard cards and analytics.
    Used by the /http-probes/stats endpoint.
    """
    total_probes: int = Field(..., description="Total number of HTTP probes")
    status_code_distribution: Dict[int, int] = Field(
        ..., 
        description="Count of probes by status code (e.g., {200: 45, 404: 10, 500: 2})"
    )
    top_technologies: List[Dict[str, Any]] = Field(
        default_factory=list,
        description="Most common technologies with counts (e.g., [{'name': 'IIS:10.0', 'count': 15}])"
    )
    top_servers: List[Dict[str, Any]] = Field(
        default_factory=list,
        description="Most common web servers with counts (e.g., [{'name': 'Apache', 'count': 20}])"
    )
    cdn_usage: Dict[str, int] = Field(
        default_factory=dict,
        description="CDN providers with counts (e.g., {'Cloudflare': 30, 'Akamai': 10})"
    )
    redirect_chains_count: int = Field(
        ..., 
        description="Number of probes that had redirect chains (chain_status_codes not empty)"
    )
    
    class Config:
        json_schema_extra = {
            "example": {
                "total_probes": 80,
                "status_code_distribution": {
                    "200": 65,
                    "404": 10,
                    "500": 5
                },
                "top_technologies": [
                    {"name": "IIS:10.0", "count": 25},
                    {"name": "Microsoft ASP.NET", "count": 20},
                    {"name": "Apache", "count": 15}
                ],
                "top_servers": [
                    {"name": "Microsoft-IIS/10.0", "count": 30},
                    {"name": "Apache", "count": 25},
                    {"name": "nginx", "count": 20}
                ],
                "cdn_usage": {
                    "Cloudflare": 15,
                    "Akamai": 5
                },
                "redirect_chains_count": 12
            }
        }
